### ISSUE TITLE
3.7-Dataloader

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#22]
### 前言

因處理數據的程式碼太混亂和難維持，所以pytorch提供了2種data primitives
torch.utils.data.DataLoader and torch.utils.data.Dataset
允許你使用pre-loaded的資料集和你的資料集進行training
所以要將data先load進去才可以進行training等步驟

![image](https://user-images.githubusercontent.com/34474924/195294223-4bb3b9c9-e527-4334-ab98-32ae7e5856f2.png)

Dataset存data和相對應的特徵(label) => 你定義一個class去代表資料集, 包括資料的前處理
DataLoader則是將Dataset用一個可迭代的東西打包，讓你可以更方便的獲取samples

________

### Dataloader

```py
#Dataloader範例

import torchvision
from torch.utils.data import DataLoader
from torch.utils.tensorboard import SummaryWriter


test_set = torchvision.datasets.CIFAR10(root="./CIFAR10_dataset", train=False, transform=torchvision.transforms.ToTensor(), download=True)

test_loader = DataLoader(dataset=test_set, batch_size=64, shuffle=True, drop_last=False)


writer = SummaryWriter("dataloader_log")
#tensorboard --logdir=dataloader_log --port=6008
step = 0
for data in test_loader:
    imgs, targets = data
    # print(imgs.shape)
    # print(targets)
    #使用images放入批次的數據
    writer.add_images("test_data", imgs, step)
    step = step + 1

writer.close()

```

_____________

為了使每次抓取的數據都不同, 可避免local min和 overfitting, 可以抓取完之後再重洗, 再抓取

也就是使用Epoch迴圈的概念

```py
#Dataloader範例

import torchvision
from torch.utils.data import DataLoader
from torch.utils.tensorboard import SummaryWriter


test_set = torchvision.datasets.CIFAR10(root="./CIFAR10_dataset", train=False, transform=torchvision.transforms.ToTensor(), download=True)

test_loader = DataLoader(dataset=test_set, batch_size=64, shuffle=True, drop_last=False)
# test_loader = DataLoader(dataset=test_set, batch_size=64, shuffle=True, drop_last=True)


writer = SummaryWriter("dataloader_log")
#tensorboard --logdir=dataloader_log --port=6008

for epoch in range(2):
    step = 0
    for data in test_loader:
        imgs, targets = data
        # print(imgs.shape)
        # print(targets)
        #使用images放入批次的數據
        writer.add_images("Epoch:{}".format(epoch), imgs, step)
        step = step + 1

writer.close()
```
